frontend/src/lib/components/PropertyFilters/PropertyFilter.js

```diff
@@ -2,99 +2,79 @@ import React, { Component } from 'react'
 import Select from 'react-select'
 import { CloseButton, selectStyle } from '../../utils'
 import { PropertyValue } from './PropertyValue'
-import PropTypes from 'prop-types'
+import { useValues, useActions } from 'kea'
+import { propertyFilterLogic } from './propertyFilterLogic'
 
-export class PropertyFilter extends Component {
-    constructor(props) {
-        super(props)
-
-        this.state = {}
+export function PropertyFilter({ index, endpoint }) {
+    const { properties, filters } = useValues(propertyFilterLogic)
+    const { set, remove } = useActions(propertyFilterLogic)
+    let item = filters[index]
+    let key = Object.keys(item)[0] ? Object.keys(item)[0].split('__') : []
+    let value = Object.values(item)[0]
+    let operatorMap = {
+        null: 'equals',
+        is_not: "doesn't equal",
+        icontains: 'contains',
+        not_icontains: "doesn't contain",
+        gt: 'greater than',
+        lt: 'lower than',
     }
-    render() {
-        let { properties, index, item, onSet, onRemove, endpoint } = this.props
-        let key = Object.keys(item)[0] ? Object.keys(item)[0].split('__') : []
-        let value = Object.values(item)[0]
-        let operatorMap = {
-            null: 'equals',
-            is_not: "doesn't equal",
-            icontains: 'contains',
-            not_icontains: "doesn't contain",
-            gt: 'greater than',
-            lt: 'lower than',
-        }
-        return (
-            <div className="row" style={{ margin: '0.5rem -15px' }}>
-                <div className="col-3" style={{ paddingRight: 0 }}>
+    return (
+        <div className="row" style={{ margin: '0.5rem -15px' }}>
+            <div className="col-3" style={{ paddingRight: 0 }}>
+                {properties && (
```

Comment:
@mariusandra I have a weird problem with this `<Select>`. I converted this component to a function with Kea. When I select an item and onChange gets fired, it takes a good 4-5 seconds before this whole component gets re-rendered. Any ideas why that might be?

Reply:
Hi, the problem is here: https://github.com/PostHog/posthog/pull/588/files/0c48d2aa460631a1bb8f78370db51ed4da5fba9b#diff-660e60cc27584640c208d462d3d1e531R50-R54

The reducer for `filters -> actions.set` and also some of the others mutate the array instead of returning a new array. React-redux uses referential checks (`newState.filters !== oldState.filters`) to check if something changed or not... and re-renders the component only if the reference to the object changed.

If you just do `state.filters.push('bla')`, then `newState.filters === oldState.filters` and nothing will update... except if something triggers the entire scene to re-render.

Try something like this:

```javascript

        filters: [
            [],
            {
                [actions.setInitial]: (_, { filters }) => {
                    if (!filters) return []
                    return Object.entries(filters).map(([key, value]) => {
                        return { [key]: value }
                    })
                },
                [actions.set]: (filters, { index, key, value }) => {
                    const newFilters = [...filters]
                    newFilters[index] = { [key]: value }
                    return newFilters
                },
                [actions.newFilter]: filters => {
                    return [...filters, {}]
                },
                [actions.remove]: (filters, { index }) => {
                    return filters.filter((f, i) => i !== index)
                },
            },
        ],
```

Reply:
Doh, that's fairly obvious with hindsight. Thanks!